### PR TITLE
Day's work 10/19/2018 - 10/20/2018:)

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
Corrected:
* kibana now seems to at least install and start (according to ambari) in the mpack
* Tests seem healthier, but I had to remove the `@{argLine}` call due to that travis error you showed me earlier.  You can recreate the travis error by reverting the removal of `@{argLine}` from the top level pom's surefire plugin config and running `mvn -q -T 2C surefire:test@unit-tests`  This evidently works differently than running the tests via `mvn clean install`
* Downgraded version of log4j from `metron-elasticsearch` as it works poorly with Storm 1.0.x (classpath issues when starting the indexing topology).

My computer literally crashed spinning up full-dev, so I haven't gotten a chance to see data flow through, but I got all the topologies starting.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [ ] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [ ] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [ ] Have you included steps or a guide to how the change may be verified and tested manually?
- [ ] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by cestella** [comment]
I added a relocation to avoid the following exception in the indexing topology:
```
java.lang.NoSuchMethodError: org.apache.logging.log4j.Logger.debug(Ljava/lang/String;Ljava/lang/Object;)V
	at org.elasticsearch.threadpool.ThreadPool.<init>(ThreadPool.java:203) ~[stormjar.jar:?]
	at org.elasticsearch.client.transport.TransportClient.buildTemplate(TransportClient.java:129) ~[stormjar.jar:?]
	at org.elasticsearch.client.transport.TransportClient.<init>(TransportClient.java:265) ~[stormjar.jar:?]
	at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:130) ~[stormjar.jar:?]
	at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:116) ~[stormjar.jar:?]
	at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:106) ~[stormjar.jar:?]
	at org.apache.metron.elasticsearch.utils.ElasticsearchUtils.getClient(ElasticsearchUtils.java:71) ~[stormjar.jar:?]
	at org.apache.metron.elasticsearch.writer.ElasticsearchWriter.init(ElasticsearchWriter.java:60) ~[stormjar.jar:?]
	at org.apache.metron.writer.bolt.BulkMessageWriterBolt.prepare(BulkMessageWriterBolt.java:189) ~[stormjar.jar:?]
	at org.apache.storm.daemon.executor$fn__6573$fn__6586.invoke(executor.clj:798) ~[storm-core-1.0.1.2.5.3.0-37.jar:1.0.1.2.5.3.0-37]
	at org.apache.storm.util$async_loop$fn__554.invoke(util.clj:482) [storm-core-1.0.1.2.5.3.0-37.jar:1.0.1.2.5.3.0-37]
	at clojure.lang.AFn.run(AFn.java:22) [clojure-1.7.0.jar:?]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_77]
```

This may not actually fix this, but the core issue is that the ES transport client requires a log4j version greater than the one mandated by storm.  If this doesn't work, we might consider seeing if storm can function with a newer log4j by excluding the older version from storm and replacing it with the log4j required in `elasticsearch-shaded`

**Comment by cestella** [comment]
Also please note that the following is failing in ambari:
`curl -s -XPOST http://node1:9200/_template/bro_index -d @/var/lib/ambari-agent/tmp/bro_index.template` seems to be failing.  I haven't had a chance to look at that yet.

**Comment by cestella** [comment]
Also found this when starting kibana:
```
No handlers could be found for logger "elasticsearch"
Traceback (most recent call last):
  File "/var/lib/ambari-agent/cache/common-services/KIBANA/5.6.2/package/scripts/kibana_master.py", line 128, in <module>
    Kibana().execute()
  File "/usr/lib/python2.6/site-packages/resource_management/libraries/script/script.py", line 280, in execute
    method(env)
  File "/usr/lib/python2.6/site-packages/ambari_commons/os_family_impl.py", line 89, in thunk
    return fn(*args, **kwargs)
  File "/var/lib/ambari-agent/cache/common-services/KIBANA/5.6.2/package/scripts/kibana_master.py", line 120, in load_template
    di.es.indices.delete(index='.kibana', ignore=[400, 404])
  File "/usr/lib/python2.6/site-packages/elasticsearch/client/utils.py", line 69, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/usr/lib/python2.6/site-packages/elasticsearch/client/indices.py", line 198, in delete
    params=params)
  File "/usr/lib/python2.6/site-packages/elasticsearch/transport.py", line 307, in perform_request
    status, headers, data = connection.perform_request(method, url, params, body, ignore=ignore, timeout=timeout)
  File "/usr/lib/python2.6/site-packages/elasticsearch/connection/http_urllib3.py", line 89, in perform_request
    raise ConnectionError('N/A', str(e), e)
elasticsearch.exceptions.ConnectionError: ConnectionError(('Connection aborted.', error(111, 'Connection refused'))) caused by: ProtocolError(('Connection aborted.', error(111, 'Connection refused')))
```

**Comment by cestella** [comment]
We need to figure out what the heck to do about the following params in elasticsearch:
* `indices.memory.store.throttle.type` to renamed to `indices.store.throttle.type`.  Is this acceptable?  I can't find a damned thing about the first param.
* `indices.cluster.send_refresh_mapping` seems to have gone away.  Is there a replacement?  Again, can't figure out where it might've gone in 5.6
* `transient.indices.store.throttle.type` seems to not be a thing anymore either.

Of all of these, the params that concern me the most are the ones around the throttle type.  I think the following is going on, though.  In ES 2.x, we are making it so that the memory store throttle is set to `none` by default and the indices.store.throttle.type is set to merge.  I am afraid that my change may have made the default throttle type set to none, which would optimize for writes over searches (see [here](https://www.elastic.co/guide/en/elasticsearch/guide/current/indexing-performance.html) ).  We probably want to get rid of the memory indices.store.throttle.type parameter and replace it with just the `indices.store.throttle.type` param.  I *think* this is the only knob they have in 5.x around that.

TL;DR:
* Don't know what `indices.cluster.send_refresh_mapping` does, so I removed it.
* Only setting `indices.store.throttle.type`, but I fear we're setting it to `none` by default when it should be `merge`.

**Comment by cestella** [comment]
I determined that we need the following in `/etc/security/limits.conf`:
```
elasticsearch - nproc 2048
elasticsearch soft memlock unlimited
elasticsearch hard memlock unlimited
```

We need to add this to ansible; if so, then things start ok.

**Comment by cestella** [comment]
With the previous commits, data flows through the indexing topology into ES5.  There are issues, however, with the templates.  Specifically the following exceptions are seen in the elasticsearch logs:
```
Caused by: java.lang.IllegalArgumentException: [_timestamp] is removed in 5.0. As a replacement, you can use an ingest pipeline to add a field with the current timestamp to your documents.
java.lang.IllegalArgumentException: mapper [version] cannot be changed from type [long] to [integer]
org.elasticsearch.index.mapper.MapperParsingException: Failed to parse mapping [bro_doc]: [_timestamp] is removed in 5.0. As a replacement, you can use an ingest pipeline to add a field with the current timestamp to your documents.
org.elasticsearch.index.mapper.MapperParsingException: Failed to parse mapping [error_doc]: [_timestamp] is removed in 5.0. As a replacement, you can use an ingest pipeline to add a field with the current timestamp to your documents.
org.elasticsearch.index.mapper.MapperParsingException: Failed to parse mapping [snort_doc]: [_timestamp] is removed in 5.0. As a replacement, you can use an ingest pipeline to add a field with the current timestamp to your documents.
org.elasticsearch.index.mapper.StrictDynamicMappingException: mapping set to strict, dynamic introduction of [hits] within [dashboard] is not allowed
org.elasticsearch.index.mapper.StrictDynamicMappingException: mapping set to strict, dynamic introduction of [sort] within [search] is not allowed
```

I suspect removing the `_timestamp` portion and ensuring the `timestamp` field works will do wonders here.

**Comment by cestella** [comment]
Also, the REST service fails to start due to a classpath issue:
```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.apache.metron.indexing.dao.IndexDao]: Factory method 'indexDao' threw exception; nested excep
tion is java.lang.NoSuchMethodError: io.netty.buffer.CompositeByteBuf.addComponents(ZLjava/lang/Iterable;)Lio/netty/buffer/CompositeByteBuf;
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
        ... 45 more
Caused by: java.lang.NoSuchMethodError: io.netty.buffer.CompositeByteBuf.addComponents(ZLjava/lang/Iterable;)Lio/netty/buffer/CompositeByteBuf;
        at org.elasticsearch.transport.netty4.Netty4Utils.toByteBuf(Netty4Utils.java:117)
        at org.elasticsearch.transport.netty4.Netty4Transport.sendMessage(Netty4Transport.java:395)
        at org.elasticsearch.transport.netty4.Netty4Transport.sendMessage(Netty4Transport.java:94)
        at org.elasticsearch.transport.TcpTransport.internalSendMessage(TcpTransport.java:1111)
        at org.elasticsearch.transport.TcpTransport.sendRequestToChannel(TcpTransport.java:1093)
        at org.elasticsearch.transport.TcpTransport.executeHandshake(TcpTransport.java:1600)
        at org.elasticsearch.transport.TcpTransport.openConnection(TcpTransport.java:568)
        at org.elasticsearch.transport.TcpTransport.openConnection(TcpTransport.java:114)
        at org.elasticsearch.transport.TransportService.openConnection(TransportService.java:351)
        at org.elasticsearch.client.transport.TransportClientNodesService$SimpleNodeSampler.doSample(TransportClientNodesService.java:407)
        at org.elasticsearch.client.transport.TransportClientNodesService$NodeSampler.sample(TransportClientNodesService.java:357)
        at org.elasticsearch.client.transport.TransportClientNodesService.addTransportAddresses(TransportClientNodesService.java:198)
        at org.elasticsearch.client.transport.TransportClient.addTransportAddress(TransportClient.java:319)
        at org.apache.metron.elasticsearch.utils.ElasticsearchUtils.getClient(ElasticsearchUtils.java:73)
```
